### PR TITLE
[ContentCacheManager] write streams to disk atomically

### PR DESCRIPTION
**Pull request by das7pad**
### Description

For https://github.com/overleaf/issues/issues/3871

This PR is making the writing of streams to disk _atomically_ using an intermediate file for writing and a final rename operation.

#### Related Issues / PRs

For https://github.com/overleaf/issues/issues/3871

### Review

In case the writing or rename operation fails, unlink the file.

#### Potential Impact

Low.

#### Manual Testing Performed

- clear output dir
- re-compile
- see cached assets

### CODE REVIEW

**Review by briangough — COMMENTED** [review]
At line 30 we should keep track of files already written and skip subsequent writes for files with the same hash.  Currently if a pdf file contains the same image multiple times we will write the file more than once, which is unnecessary.  This can happen if the user has uploaded the same file under different names, since pdftex doesn't recognise them as being the same in that case and puts multiple objects into the pdf file.

Example of logging the hashes from a project with the same image uploaded with two different file names:
```
\includegraphics[width=\textwidth]{aerials/2-2-19.png}
\includegraphics[width=\textwidth]{aerials/2-2-19a.png}  % same image, different filename
```

```
clsi_1                    | writing d78ea42cc2d44ba1660625f5dd9c9c9435c2c818b212fffc27b3a9f49fcb8fbf
clsi_1                    | writing d78ea42cc2d44ba1660625f5dd9c9c9435c2c818b212fffc27b3a9f49fcb8fbf
```

File: app/js/ContentCacheManager.js [hunk 1]
```diff
@@ -100,17 +100,27 @@ async function writePdfStream(dir, hash, buffers) {
     //  ETags used for client side caching via browser internals.
     return false
   } catch (e) {}
-  const file = await fs.promises.open(filename, 'w')
+  const atomicWriteFilename = filename + '~'
+  const file = await fs.promises.open(atomicWriteFilename, 'w')
```
[the remark this answers is not in the document]
Comment (das7pad) [reply app/js/ContentCacheManager.js:104]: There is a lock around the compile operation, so we will have at most writer. The flushing of ranges is sequential as well. Do we need to add the exclusive flag then?

**Comment by das7pad** [comment]
> At line 30 we should keep track of files already written and skip subsequent writes for files with the same hash.

Good idea, let's do that in a flow-up PR? It seems to be an optimization, whereas this PR is rather a bug fix.

File: app/js/ContentCacheManager.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (das7pad) [reply app/js/ContentCacheManager.js:104]: If we go down that route, we probably need to add another try/catch around this call then and finger printing on the error for ignoring the case of already existing file. This seems like complexity we can skip, WDYT?

[the remark this answers is not in the document]
Comment (briangough) [reply app/js/ContentCacheManager.js:104]: It's a pity node doesn't support `mkstemp`.

**Comment by briangough** [comment]
> > At line 30 we should keep track of files already written and skip subsequent writes for files with the same hash.
> 
> Good idea, let's do that in a flow-up PR? It seems to be an optimization, whereas this PR is rather a bug fix.

That is fine.

**Review by briangough — APPROVED** [review]

File: app/js/ContentCacheManager.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (briangough) [reply app/js/ContentCacheManager.js:104]: I think it's ok to leave it as is, it is just defensive.

**Pull request merged**